### PR TITLE
Fixed suite TearDown ordering 

### LINKED ISSUE #466: bug in parallel tests
When enable parallel testing, we would expect the TearDownSuite() to be called after all the tests completed. However, after this commit, https://github.com/stretchr/testify/commit/3104bf5483d3fba84f24933949f0533f4467880b
the parallel testing is broken as the TearDownSuite() is called before tests are done.
Here is a simple repro code to repro:
```
type MyTestSuite struct {
	suite.Suite
}
func TestMySuite(t *testing.T) {
	suite.Run(t, new(MyTestSuite))
}
func (s *MyTestSuite) SetupSuite() {
	fmt.Println("SetupSuite")
}
func (s *MyTestSuite) TearDownSuite() {
	fmt.Println("TearDownSuite")
}
func (s *MyTestSuite) SetupTest() {
	s.T().Parallel()
	fmt.Println("SetupTest")
}
func (s *MyTestSuite) TearDownTest() {
	fmt.Println("TearDownTest")
}
func (s *MyTestSuite) Test_A() {
	time.Sleep(time.Second)
}
func (s *MyTestSuite) Test_B() {
	time.Sleep(time.Second * 2)
}

```
The output of above tests is:
```
SetupSuite
TearDownSuite
SetupTest
SetupTest
TearDownTest
TearDownTest
```

### PR DESCRIPTION
**Pull request by devdinu**
Using WaitGroup to synchronise between tests so TearDown runs at end.

Having `suite.T().Parallel()` in `TestA` causes it to hang. goroutine wait in `<-t.signal` at [testing.go](https://github.com/golang/go/blob/master/src/testing/testing.go#L961) and suite.go Run to wait on `testsSync` `:108` 

Closes #466

### CODE REVIEW

**Comment by devdinu** [comment]
@georgelesica-wf could we merge this.

**Comment by georgelesica-wf** [comment]
CI is broken right now and I haven't had time to fix it. It's not running our checks, just the tests. If someone can figure out how to get the code gen to run and can finish making [this PR](https://github.com/stretchr/testify/pull/804) pass then we can start merging again. Until then I'd rather hold off.

**Comment by devdinu** [comment]
@boyan-soubachov Could you help reviewing this and merge.

File: suite/suite_order_test.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package suite
```
Comment (boyan-soubachov) [inline suite/suite_order_test.go]: What is your logic for putting this in a separate file? Does it not make sense to just have this as another test in `suite_test.go`?

File: suite/suite.go [hunk 2]
```diff
@@ -111,6 +114,9 @@ func Run(t *testing.T, suite TestingSuite) {
 		test := testing.InternalTest{
 			Name: method.Name,
 			F: func(t *testing.T) {
+				defer func() {
```
Comment (boyan-soubachov) [inline suite/suite.go]: Can we not just `defer testSync.Done()` instead of wrapping it in an anonymous function?

Comment (devdinu) [reply suite/suite.go]: Had some logic which i've removed later. good catch, have updated.

File: suite/suite_order_test.go (hunk 1 above)
Comment (devdinu) [reply suite/suite_order_test.go]: The other file becoming huge, thought could separate the tests based on what we integration vs small func tests.

Comment (boyan-soubachov) [reply suite/suite_order_test.go]: I see. I do agree that it's becoming a bit large but I don't think navigating it is still a huge problem. 

I'd be happy to discuss any ideas you might have about improving its readability & maintainability, whether it involves splitting it up or some other way.

**Pull request merged**